### ISSUE TITLE
has_parent gaps in 3.0.227 (was ok in 3.0.206)

### ISSUE THREAD

**Issue by er314** [processwire/processwire-issues#1814]
### Short description of the issue

Under specific scenario involving page move, pages()->find selector with 'has_parent' condition is failing to retrieve the full list of pages having the requested parent.
This is with PW 3.0.227.
It used to be ok with PW 3.0.206.

### Expected behavior

Return all pages having the requested parent.

### Steps to reproduce the issue

In the below scenario, 
- Before the page move, it is expected that the selector returns 3 pages having the requested parent 
-> OK
- After the page move, it is expected that the selector returns 4 pages having the requested parent
-> NOK, it returns 2 pages instead. 
(while it returns the expected 4 pages in PW 3.0.206)
It should be noted that this is fixed as soon as "hasParent_rebuildAll()" is executed.
So this is clearly a bug in PW handling of hasParent.
It should also be noted that I also experienced a similar situation when moving page via the GUI (hard to say if it's the exact same issue or not).

This is the scenario :

```
	public function test_Processwire_has_parent_2(): void 
	{
		$desc = "test Processwire-core, has_parent #2";

		$rootPath = "/data/unit-test/processwire-core-tests/subdir1/";
		$rootPg = pages()->get($rootPath);

		// create main page
		$mainPage = new Page();
		$mainPage->template = 'basic-page';
		$mainPage->parent = $rootPath;
		$mainPage->title = "test main page";
		$mainPage->of(false);
		$mainPage->save();

		// create subfolder page
		$folderPage = $mainPage->child("title=folder-page, include=all");
		if ($folderPage instanceof NullPage) {
			$folderPage = new Page();
			$folderPage->template = 'basic-page';
			$folderPage->parent = $mainPage;
			$folderPage->of(false);
			$folderPage->title = "test subfolder";
			$folderPage->save();
			$folderPage->of(true);
		}

		// create leaf page
		$leafPage = new Page();
		$leafPage->of(false);
		$leafPage->template = 'basic-page';
		$leafPage->parent = $folderPage;
		$leafPage->title = "test leaf page";

		$leafPage->save();
		$leafPage->of(true);

		$arrPgHasParent = pages()->find("has_parent={$rootPath}, include=all");
		$actual = count($arrPgHasParent);
		$expected = 3;  // = (rootPath/) mainPage / folderPage / leafPage
		$this->assertEquals($expected, $actual, "{$desc} 1");

		// move the page to another root, with ad-hoc creation of subfolder
		$rootPathDest = "/data/unit-test/processwire-core-tests/subdir2/";
		$rootPgDest = pages()->get($rootPathDest);

		// create new "sub-root" subfolder
		$targetFolder = new Page();
		$targetFolder->template = 'basic-page';
		$targetFolder->parent = $rootPgDest;
		$targetFolder->of(false);
		$targetFolder->title = "newSubdir";
		$targetFolder->save();
		$targetFolder->of(true);

		// perform the move
		$currentFolder = $mainPage->parent();
		if ($currentFolder->id !== $targetFolder->id) {
			$mainPage->of(false);
			$mainPage->parent = $targetFolder;
			$mainPage->save();
			$mainPage->of(true);
		}

		$arrPgHasParent = pages()->find("has_parent={$rootPathDest}, include=all");
		$actual = count($arrPgHasParent);
		$expected = 4;  // = (rootPathDest/) targetFolder / mainPage / folderPage / leafPage
		$this->assertEquals($expected, $actual, "{$desc} 2");

		// cleanup
		$res = PwPage::forceDeleteRecursive($rootPgDest, isIncludeRoot: false);
	}

```

### Follow-up

Can you please seriously consider releasing the currently closed-source tests of the selector engine, as open source unit-tests, accompanying Processwire source code ?
This would allow us (the community) to contribute to it and enrich it via PR.
Instead of the current odd situation where I (and others) are adding Processwire-core dedicated non-regression unit-tests to our own apps test cases. 
Like this test case -> it would benefit from being added to PW unit-tests.

**Comment by ryancramerdesign** [comment]
@er314 I didn't understand the example, as it looks like this might be part of a larger class that I don't have the full context for. Though I'm not sure I follow well enough to reproduce that example on the code side either way, so I tried to reproduce what you describe interactively instead. After experimenting a bit, I was able to duplicate what I think you are seeing, and have added a fix for it. Though I'd be curious if it also fixes it for you. If not, try doing a `$pages->parents()->rebuildAll()` once, and then try again, just in case it needs a reset to get things in order. 

In the 3.0.206 version you describe, ProcessWire basically did the rebuildAll() on every page movement. That's usually pretty fast, but it turned out that on some really large sites, it could be incredibly slow too. So the updated code now is more of a scalpel to just focus in on what needs to be updated. But the original rebuildAll() code is still there, it's just not used for everything like it was before. 

I don't know about these closed source tests you mentioned, at least I do not maintain that. Whenever I work on the core and add or change anything that needs testing, I build various specific tests to verify they do what they are supposed to do. But I don't keep those tests around in a library or anything like that. Though a real test framework is definitely on the roadmap.

**Comment by er314** [comment]
> I'd be curious if it also fixes it for you

It's better, but there is still one gap ! (one missing page)
Expectation is to return 4 pages, before this fix it was 2, now it's 3.
**--> In the scenario provided, the missing page is "test main page".**

Yep, rebuildAll() is returning the 4 pages.


> it looks like this might be part of a larger class that I don't have the full context for

Well, the class is using PHPUnit (https://phpunit.de/),
```
use PHPUnit\Framework\TestCase;
class ProcesswireTest extends TestCase {
```
which gives the capacity to do the test assertion :  
```
$arrPgHasParent = pages()->find("has_parent={$rootPathDest}, include=all");
$actual = count($arrPgHasParent);
$expected = 4;
$this->assertEquals($expected, $actual);
```
Apart from the assertion instructions, the rest of the code is (should be) 100% ProcessWire native.
So basically, the code is : 
1. constructing this page tree : 
`rootPath / mainPage / folderPage / leafPage`
2. constructing this page tree : 
`rootPathDest / targetFolder /`
3. moving "mainPage" tree below "targetFolder"
So the result of the move is :  
`rootPathDest / targetFolder / mainPage / folderPage / leafPage`
And when querying for the pages having "rootPathDest" as parent, we expect to get the 4 pages below it.
But we get only 3 ; "mainPage" is still missing.

> In the 3.0.206 version you describe, ProcessWire basically did the rebuildAll() on every page movement. 
[...] the updated code now is more of a scalpel

Okaayy ! so the updated code is definitely the better approach, but this change explains that some bugs can arise.

> I don't know about these closed source tests you mentioned

This :-) : https://github.com/niklaka/ProcessWireTests/pull/2
plus from memory, some ancient forums talks involving Pete or Anti.
Anyway, all of this surely revolves around the current state, which I call "closed source" in that it's kept on your side : _"Whenever I work on the core and add or change anything that needs testing, I build various specific tests to verify they do what they are supposed to do"._

> a real test framework is definitely on the roadmap

This is great news. For many people (https://github.com/processwire/processwire-requests/issues/490).

**Comment by ryancramerdesign** [comment]
Thanks @er314 I had a little trouble duplicating this, but finally did. Then I could duplicate it once, and the issue fixed itself after another move. Finally, I found one parent that I could consistently duplicate it with, while strangely not with another, having the same structure. What I found is that because the function queries its own table (pages_parents) to get information, if there's something missing in it (perhaps due to a prior error), then it'll continue to manifest in further updates, at least in some conditions. So I added some additional code that also queries the pages directly for parent information, as an extra redundancy and verification. So if there's some previous missing parent in the table, then it'll get fixed automatically on the next move. At least that's the intention. It seems to be working here, but I'm curious if you find it also resolves it there? My test case where I could duplicate it ended up being a little different than yours.
